README.md
# notOS

> Web-based fake operating system
> 
> https://www.notkal.com/notOS/

**To-do:**
- [ ] Accept dragged images as a way to create an image file (?)
- [ ] Backdrop blur effect not on mobile


changelog.txt
v1.2.3
- Gave files app a new icon
- Gave folders a new icon
- Fixed the desktop being scrollable when not necessary

v1.2.2
- Apps now have the .app file extension and can be double click opened
- Apps in file views now use their icons as a thumbnail
- Amount of files in a folder now shows at the bottom of files

v1.2.1
- Directory now appears at the bottom of Files

v1.2
- General updates and fixes
- Added web browser
- Clicking a window will now bring it to the front
- Fixed dock icons not appearing

v1.1
- Images now use themselves as their thumbnail

v1
Desktop, Dock, Folders
Notes, Images, Files and, Dialog apps

main.js
// Shorthand
function style(element, classname, state) { state ? element.classList.add(classname) : element.classList.remove(classname); }


var pX = 0;
var pY = 0;

// DOM
const container = document.getElementById('app_container');

// Variables
const process_template = {
    memory: {}, // data

    max: false,
    min: false,

    width: '400',
    height: '600',
    x: 48,
    y: 48,
    z: 1,
    allow_actions: true,

    app: 'notes.app',
    id: 1,
    element: undefined,
}
var processes = {}
var pid = 1;

var temp_z = 1;

var dragging = false;
let dragOffset = {
    x: 0,
    y: 0,
}

/** Takes directory string an returns file */
function getFile(file) {
    let target = file_system;

    if(typeof file == 'string') {
        let dir = file.split('/');
        dir.shift();
        for(i in dir) {
            let f = dir[i];
            // console.log(f);
            target = target[f];
        }
    } else target = { data: '', type: 'temp', }
    return target;
}

/** Launches an application */
function launch(name='notes.app', file, open_existing=false) {
    // Existing process
    console.log(open_existing);
    // if(open_existing == true) {
    //     let dockicon = document.querySelector('#dock .dock_app[data-process]');
    //     if(dockicon != null) {
    //         let proc = processes[dockicon.dataset.process];
    //         console.log(proc);
    //         return;
    //     }
    // }

    let app = file_system.apps[name];
    if(app == undefined) return console.warn('Application does not exist');

    // Create process
    let proc = processes[pid] = JSON.parse(JSON.stringify(process_template));
    proc.app = name;
    proc.id = pid;
    proc.width = app.window.width;
    proc.height = app.window.height;
    proc.allow_actions = app.window.allow_actions;
    // let offset = (Object.keys(processes).length || 1) * 24;
    // console.log(offset);
    // proc.x = offset;
    // proc.y = offset;

    // HTML
    var wrapper = document.createElement('div');
    wrapper.innerHTML = appHTML(app, proc);
    container.append(wrapper.firstElementChild);
    proc.element = document.querySelector(`#app_container .window[data-process="${proc.id}"]`);

    // Temporary Z-Index
    proc.element.addEventListener('click', () => {proc.element.style.zIndex = temp_z; temp_z++;})
    proc.element.style.zIndex = temp_z; temp_z++;

    // Title bar
    let titlebar = document.querySelector(`#app_container .window[data-process="${proc.id}"] > .title_bar`);
    titlebar.addEventListener('mousedown', event => {
        dragging = proc.id;
        // console.log(event.target.tagName);
        if(event.target.tagName == 'BUTTON') return;
        dragWindow(event, app, proc);
    });
    titlebar.addEventListener('dblclick', event => {
        if(event.target.tagName == 'BUTTON') return;
        windowAction('max', proc.id);
    })

    // Update dock
    let dockicon = document.getElementById(`dock_${name}`);
    if(dockicon != null) {
        dockicon.classList.add('active');
        dockicon.dataset.process = proc.id;
    };

    // Script
    if(app.script != undefined) app.script('start', proc, file);

    pid++; // Increment
    /** Template */
    function appHTML(app, proc) {
        let icon = app.icon == false ? '' : `<img src="${app.icon}" alt="" class="title_icon">`;
        let name_raw = proc.app.split('.')[0];
        let title_buttons = proc.allow_actions == false ? '' : `
        <button onclick="windowAction('min', ${proc.id})">_</button>
        <button onclick="windowAction('max', ${proc.id})">[]</button>
        <button class="button_close" onclick="windowAction('close', ${proc.id})">X</button>`;
        return `
        <div class="window app_${name_raw}" data-process="${proc.id}" data-app="${name_raw}" data-open-file="~/desktop/Example.txt" style="width: ${proc.width}px; height: ${proc.height}px; top: ${proc.y}; left: ${proc.x}; z-index: ${proc.z};">
            <!-- Title bar -->
            <div class="title_bar">
                ${icon}
                <h2>${app.title}</h2>
                ${title_buttons}
            </div>

            <div class="content">
                ${app.html}
            </div>
        </div>`;
    }
}

/** Close/max/min window */
function windowAction(action, id) {
    if(action == undefined || id == undefined) return console.warn('Invalid window action');
    let proc = processes[id];
    let app = file_system.apps[proc.app];
    if(proc == undefined) return console.warn(`Process (${id}) does not exist`);
    // if(proc.allow_actions == false) return console.warn('This window has disabled this feature');

    // Action
    if(action == 'close') {
        proc.element.classList.add('window_close');
        proc.close_timer = setTimeout(() => {
            proc.element.remove();
            delete processes[proc.id];
        }, 100);
        let dockicon = document.getElementById(`dock_${proc.app}`);
        if(dockicon != null) dockicon.classList.remove('active');
    }
    else if(action == 'max') {
        proc.max = !proc.max;
        style(proc.element, 'state_max', proc.max);
    }
    else if(action == 'min') {
        proc.min = !proc.min;
        style(proc.element, 'state_min', proc.min);
    }
    else console.warn('Invalid window action');
}

/** Drag window */
function dragWindow(event, app, proc) {
    if(dragging == false) return;
    if(proc.max) windowAction('max', proc.id);

    if(event != false) {
        let rect = event.target.getBoundingClientRect();
        dragOffset.x = event.clientX - rect.left;
        dragOffset.y = event.clientY - rect.top;
    }

    let [x, y] = [pX - dragOffset.x, pY - dragOffset.y];

    proc.x = x;
    proc.y = y;
    proc.element.style.left = `${x}px`;
    proc.element.style.top = `${y}px`;

    requestAnimationFrame(() => { dragWindow(false, app, proc); });
}

/** Toggles a system menu or overlay */
function menu(id) {

}

/** Opens a file */
function openFile(event) {
    let element = event.srcElement;
    if(element.tagName != 'FIGURE') element = element.parentNode;
    // console.log(element);

    let path = element.dataset.directory;
    if(isFolder(path)) {
        let proc = processes[element.dataset.process];
        proc.memory.navigate(path);
    };
    let [leftover, file_ext] = path.split('.');
    if(typeof file_ext != 'string') return console.warn('Invalid file extension');

    // .app
    console.log(leftover.split('/')[2]);
    if(file_ext == 'app') return launch(`${leftover.split('/')[2]}.app`);

    // Normal file
    let app = fileAssociation(path);
    launch(app, path);
}

/** Takes a path or filename and returns the associated app */
function fileAssociation(path) {
    let file_ext = path.split('.')[1];
    return file_system.system.config.file_association[file_ext];
}


/** Populate dock */
function populateDock() {
    let pinned = file_system.system.config.dock.pinned;
    let html = '';
    for(i in pinned) {
        let name = pinned[i];
        let app = file_system.apps[name];
        if(app == undefined) { console.warn('Dock: app doesnt exist'); continue; }
        html += `
        <button id="dock_${name}" class="dock_app" onclick="launch('${name}', undefined, true)">
            <img src="${app.icon}" alt="">
            <div class="dock_indicator"></div>
        </button>`;
    }

    document.getElementById('dock_apps').innerHTML = html;
}

function isFolder(file) { if(typeof file == 'string') file = getFile(file); return file.type == 'dir'; }

/** Populate folder */
function populateFolder(path='~/desktop', proc={id:-1}) {
    let folder = getFile(path);
    if(!isFolder(folder)) return console.warn(`(${path}) is not a folder`);
    let html = '';

    for(filename in folder) {
        let file = folder[filename];
        if(filename == 'type' && typeof file == 'string') continue;
        html += template(path, filename, file);
    }

    // Export HTML
    let containers = document.querySelectorAll(`.view_folder[data-directory="${path}"`);
    containers.forEach(element => element.innerHTML = html);

    /** Template */
    function template(path, filename, file) {
        let app = fileAssociation(filename);
        let icon;
        try {
            if(app == 'images.app') icon = file.data;
            else if(app == 'self') icon = file_system.apps[filename].icon;
            else icon = file_system.apps[app]['file_icon'];
        } catch (error) {
            if(isFolder(file)) icon = file_system.apps['files.app'].file_icon
            console.warn(error);
        }
        if(icon == undefined) icon = './assets/icon/blank_document.png';
        return `
        <figure class="file" data-process="${proc.id}" data-directory="${path}/${filename}" role="button" tabindex="0" ondblclick="openFile(event)" oncontextmenu="contextMenu(event)">
            <img src="${icon}" alt="${filename}">
            <figcaption>${filename}</figcaption>
        </figure>`;
    }

    return Object.keys(folder).length - 1;
}

const context_menu_container = document.getElementById('context_menu_container');
const context_menu = document.getElementById('context_menu');
/** Context menu */
function contextMenu(event) {
    event.preventDefault();
    let source = event.srcElement;
    console.log(source);

    let y = pY;
    let x = pX;
    let vw = Math.max(document.documentElement.clientWidth || 0, window.innerWidth || 0);
    if(x + 240 >= vw) x = vw - 240;

    // Update page
    // populateContextMenu(event.srcElement.id);
    context_menu.style.top = `${y}px`;
    context_menu.style.left = `${x}px`;
    context_menu.focus();

    /** Populate context menu */
    function populateContextMenu(id) {
        let html = '';
        let data = contextData?.[id];
        if(data == undefined) return;
        for(i = 0; i < data.length; i++) {
            console.log(data.name);
            console.log(data.code);
            html += `<div class="context_item" tabindex="0" role="button" onclick="${data.code}">${data.name}</div>`;
        }
        context_menu.innerHTML = html;
    }
}

document.addEventListener('mouseup', () => {
    let proc = processes[dragging];
    if(proc != null) {
        if(proc.y < 0) proc.y = 0;
        proc.element.style.left = `${proc.x}px`;
        proc.element.style.top = `${proc.y}px`;
    }
    dragging = false;
})

// On load
onload = () => {
    populateFolder(path='~/desktop');
    populateDock();
}

(onmousemove = event => {
    pX = event.x;
    pY = event.y;
})()


file_system.js
var file_system = {
    'desktop': {
        type: 'dir',

        'Example.txt': {
            data: 'Hello. This is a sample text file. Try editing it.',
        },
        'Another file.txt': {
            data: 'another file!!!!!!!!11!1!'
        },
        'se1.png': {
            data: './assets/wallpaper/se1.png',
        },
        'sml.txt': {
            data: 'small filename',
        },
        'really long file name.txt': {
            data: 'well',
        },
    },
    'documents': {
        type: 'dir',
        'A Document.txt': {
            data: 'Important information',
        },
        'Folder': {
            type: 'dir',
            'important folder stuff.txt': {
                data: 'very important',
            }
        }
    },
    'pictures': {
        type: 'dir',
        'se1.png': { data: './assets/wallpaper/se1.png', },
        'scr00016.png': { data: './assets/wallpaper/scr00016.png', },
        'scr00023.png': { data: './assets/wallpaper/scr00023.png', },
        'scr00030.png': { data: './assets/wallpaper/scr00030.png', },
        'scr00040.png': { data: './assets/wallpaper/scr00040.png', },
        'scr00067.png': { data: './assets/wallpaper/scr00067.png', },
        'scr00121.png': { data: './assets/wallpaper/scr00121.png', },
        'scr00142.png': { data: './assets/wallpaper/scr00142.png', },
    },
    'apps': {
        type: 'dir',

        'files.app': {
            title: 'Files',
            icon: './assets/icon/files.svg',
            file_icon: './assets/icon/folder.svg',
            html: /*html*/ `
                <style>
                    .app_files .content {
                        padding: 6px 8px;
                        display: flex;
                        flex-direction: column;
                    }
                    .app_files .content .status_bar {
                        border-top: 1px solid var(--border-color);
                        padding-top: 3px;
                        display: flex;
                        justify-content: space-between;
                        width: 100%;
                    }
                    .app_files .content .status_bar > span { flex: 1; }
                    .app_files .content .inner {
                        width: 100%;
                        height: 100%;
                        margin-bottom: 6px;

                        display: flex;
                    }
                    .app_files .content .pane {
                        width: 100%;
                        min-width: 100px;
                        max-width: 200px;
                        resize: horizontal;
                        margin-top: 12px;
                        padding-right: 12px;
                        border-right: 1px solid var(--border-color);

                    }
                    .app_files .content .pane button {
                        color: var(--text-alt);
                        font-size: 11pt;
                        text-align: left;
                        cursor: pointer;

                        width: 100%;
                        margin: 1px 6px;
                        padding: 3px 6px;
                        border-radius: 3px;
                        border-bottom: 1px solid var(--border-color);
                        transition: 0.05s background-color ease-in;
                    }
                    .app_files .content .pane button:hover {
                        color: var(--text);
                        border-color: transparent;
                        background-color: var(--hover-color);
                        transition: unset;
                    }
                    .app_files .content .pane button.active {
                        color: var(--text);
                        border-color: transparent;
                        background-color: var(--border-color);
                    }

                </style>

                <div class="options">
                    BUTTONS GO HERE
                </div>

                <div class="inner">
                    <div class="pane">
                        <button class="active" data-directory-button="~/desktop">Desktop</button>
                        <button data-directory-button="~/documents">Documents</button>
                        <button data-directory-button="~/pictures">Pictures</button>
                        <button data-directory-button="~/apps">Apps</button>
                        <button data-directory-button="~/system">System</button>
                    </div>
                    <div class="view_folder state_icon" data-directory="~/desktop">

                    </div>
                </div>
                <div class="status_bar">
                    <span class="path_name">#path</span>
                    <span class="file_count">null items</span>
                    <span class="">Stuff</span>
                </div>
            `,
            script: (stage='start', proc, path='~/desktop') => {
                // Navigate function
                proc.memory.navigate = dir => {
                    let container = document.querySelector(`#app_container .window[data-process="${proc.id}"] .view_folder`);
                    container.setAttribute("data-directory", dir);

                    let path_name = document.querySelector(`#app_container .window[data-process="${proc.id}"] .path_name`);
                    path_name.innerText = dir;

                    let file_count = document.querySelector(`#app_container .window[data-process="${proc.id}"] .file_count`);
                    file_count.innerText = `${populateFolder(dir, proc)} items`;
                }

                // Buttons
                let buttons = document.querySelectorAll(`#app_container .window[data-process="${proc.id}"] button[data-directory-button]`);
                buttons.forEach(element => {
                    element.addEventListener('click', event => {
                        try {
                            proc.memory.navigate(event.srcElement.dataset.directoryButton);
                            buttons.forEach(e => e.classList.remove('active'));
                            element.classList.add('active');
                        } catch (error) { console.warn('Invalid directory'); console.error(error);  }
                    })
                })

                proc.memory.navigate(path);
            },

            window: {
                width: '800',
                height: '500',
            },
        },

        'notes.app': {
            title: 'Notes',
            icon: './assets/icon/Bloc_Notes_SZ.png',
            file_icon: './assets/icon/blank_document.png',
            html: /*html*/ `
                <style>
                    .app_notes .content {
                        padding: 6px 8px;
                        display: flex;
                        flex-direction: column;
                    }
                    .app_notes .content textarea {
                        width: 100%;
                        height: 100%;
                        font-family: var(--font-system);
                        font-size: 11pt;
                        background: none;
                        color: var(--text);
                        border: none;
                        outline: none;
                        resize: none;
                    }
                    .app_notes .content .status_bar {
                        border-top: 1px solid var(--border-color);
                        padding-top: 3px;
                    }
                </style>

                <textarea name="" id="" cols="30" rows="10">Lorem ipsum, dolor sit amet consectetur adipisicing elit. Quis omnis quae praesentium rem qui sit laborum, iure dolor ex quibusdam illo quia aut similique, nisi voluptatibus. Ex fugit quae explicabo iusto, molestias minus? Omnis animi qui harum esse labore, quia quae aliquid, eligendi totam autem, blanditiis nihil. Aliquam, ipsa debitis?</textarea>
                <div class="status_bar">✓ Saved</div>
            `,
            script: (stage, proc, file) => {
                let target = getFile(file);

                let textarea = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content textarea`);
                let status = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content .status_bar`);
                textarea.value = target.data;

                proc.element.addEventListener('keydown', event => {
                    if(event.key == 's' && event.ctrlKey) {
                        event.preventDefault();
                        saveTXT();
                    }
                });

                textarea.addEventListener('input', () => {
                    status.innerText = 'Unsaved changes';
                });

                function saveTXT() {
                    // New file
                    if(target.type == 'temp') {
                        console.log('Create new file?');
                    }
                    target.data = textarea.value;
                    status.innerText = '✓ Saved';
                }
            },

            window: {
                width: '400',
                height: '600',
            },
        },
        'images.app': {
            title: 'Image Viewer',
            icon: './assets/icon/Antu_google-draws.svg',
            file_icon: './assets/icon/Antu_google-draws.svg',
            html: /*html*/ `
                <style>
                    .app_images .content {
                        padding: 6px 8px;
                        display: flex;
                        flex-direction: column;
                        justify-content: center;
                        overflow: hidden;
                    }
                    .app_images .content .viewer_image {
                        max-height: 100%;
                    }
                </style>

                <img src="" alt="" class="viewer_image">
            `,
            script: (stage, proc, file) => {
                let target = getFile(file);

                let image = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content .viewer_image`);
                image.src = target.data;
            },

            window: {
                width: '800',
                height: '600',
            },
        },
        'dialog.app': {
            title: 'Dialog',
            icon: false,
            html: /*html*/ `
                <style>
                    .app_dialog .content {
                        padding: 6px 8px;
                        text-align: center;
                    }
                    .app_dialog .content p {
                        padding: 16px;
                    }
                    .app_dialog .content .dialog_buttons {
                        display: flex;
                    }
                    .app_dialog .content button {
                        color: black;
                        width: 90%;
                        background-color: #cccccc;
                    }
                </style>

                <p>This is a dialog</p>
                <div class="dialog_buttons">
                    <button>Cancel</button>
                    <button>OK</button>
                </div>
            `,
            script: (proc, options) => {
                let title = document.querySelector(`#app_container .window[data-process="${proc.id}"] .title_bar h2`);
                let text = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content p`);
                title.innerText = options.title || 'Dialog';
                text.innerText = options.text || '';
            },

            window: {
                width: '400',
                height: '140',

                allow_actions: false,
            }
        },

        'browser.app': {
            title: 'Browser',
            icon: './assets/icon/moon.svg',
            file_icon: './assets/icon/Antu_google-draws.svg',
            html: /*html*/ `
                <style>
                    .app_browser .content iframe {
                        width: 100%;
                        height: 100%;
                    }
                    .app_browser .content .browser_bar {
                        display: flex;
                    }
                    .app_browser .content .url_bar {
                        width: 100%;
                        padding: 8px 12px;
                        background-color: unset;
                        border: none;
                        color: var(--text);
                        font-size: 11pt;
                    }
                </style>

                <div class="browser_bar">
                    <!-- <button class="browser_back">Back</button>
                    <button class="browser_forward">Forward</button> -->
                    <input type="url" name="url_bar" id="url_bar" class="url_bar" value="">
                </div>
                <iframe src="" frameborder="0"></iframe>
            `,
            script: (stage, proc, url="https://www.bing.com/") => {
                let iframe = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content iframe`);
                iframe.src = url;

                let url_bar = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content input.url_bar`);
                url_bar.value = url;

                // proc.memory.back = e => iframe.contentWindow.history.back();
                // proc.memory.forward = e => iframe.contentWindow.history.forward();
                // proc.memory.go = to => iframe.src = to;

                // let browser_back = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content .browser_back`);
                // let browser_forward = document.querySelector(`#app_container .window[data-process="${proc.id}"] .content .browser_forward`);
                // browser_back.addEventListener('click', () => { proc.memory.back(); });
                // browser_forward.addEventListener('click', () => { proc.memory.forward(); });
            },

            window: {
                width: '800',
                height: '600',
            },
        },
    },
    'system': {
        type: 'dir',

        'config': {
            file_association: {
                'txt': 'notes.app',

                'img':  'images.app',
                'png':  'images.app',
                'jpg':  'images.app',
                'jpeg': 'images.app',
                'gif':  'images.app',

                'app':  'self',
            },
            dock: {
                'pinned': [
                    'files.app',
                    'browser.app',
                    'notes.app',
                    'images.app',
                ]
            },
        }
    }
}